### ISSUE TITLE
dos2unix Conversion Failure

### ISSUE THREAD

**Issue by the0neWhoKnocks** [binhex/arch-delugevpn#112]
This issue seems to only crop up when dealing with an NFS mount for the config (and a VPN is enabled). On boot of Deluge, `build/root/install.sh` runs this
```sh
# convert CRLF (windows) to LF (unix) for ovpn
/usr/bin/dos2unix "${VPN_CONFIG}" 1> /dev/null
```
and this is the result
```sh
dos2unix: Failed to change the owner and group of temporary output file /config/openvpn/d2utmpKFxiOs: Invalid argument
```
Which seems to be related to https://access.redhat.com/solutions/3168671.

A possible solution would be to duplicate the file, and use the normalized version instead.
```sh
dos2unix -n "/config/openvpn/$VPN_CONFIG" /tmp/converted.ovpn
export VPN_CONFIG=/tmp/converted.ovpn
```

Or better yet, do a check to see if the file needs to be converted to begin with.
```sh
containsCRLF=$(find "/config/openvpn/$VPN_CONFIG" -not -type d -exec file "{}" ";" | grep CRLF)
if [ -z "$containsCRLF" ]; then
  /usr/bin/dos2unix "${VPN_CONFIG}" 1> /dev/null
fi

# or a combined solution
if [ -z "$containsCRLF" ]; then
  dos2unix -n "/config/openvpn/$VPN_CONFIG" /tmp/converted.ovpn
  export VPN_CONFIG=/tmp/converted.ovpn
fi
```

**Comment by the0neWhoKnocks** [comment]
Ok, got some sleep and had some time to kick the tires on my pseudo code. Once I figured out that you're compiling the `init.sh` script from multiple shell scripts, I downloaded `init.sh` from a running Container and swapped
```sh
# convert CRLF (windows) to LF (unix) for ovpn
/usr/bin/dos2unix "${VPN_CONFIG}" 1> /dev/null
```
for
```sh
set +e
containsCRLF=$(find "${VPN_CONFIG}" -not -type d -exec file "{}" ";" | grep CRLF)
set -e
if [[ ! -z "${containsCRLF}" ]]; then
  dos2unix -n "${VPN_CONFIG}" /tmp/converted.ovpn
  export VPN_CONFIG=/tmp/converted.ovpn
  echo "[info] The provided OpenVPN config contained CRLF line endings, so it was converted" | ts '%Y-%m-%d %H:%M:%.S'
fi
```

I tested it with a CRLF & LF file - it used the `converted` file when it encountered CRLF, and did nothing for the LF file.

**Comment by binhex** [comment]
Thanks for the details, I will take a look at this when I get back from my break.

**Comment by binhex** [comment]
ive decided to write a wrapper script for this, as its used in several places for different file types, i need to test it out later, but it looks like it should do the job, see what you think:-
https://github.com/binhex/scripts/blob/master/shell/arch/docker/dos2unix-wrapper.sh

**Comment by the0neWhoKnocks** [comment]
Part of the issue was that you're trying to alter a file that may not have the NFS permissions to be altered. Hence the reassignment of `VPN_CONFIG` to the converted file within the Container. It looks like you're trying to do the same thing with `mv -f "/tmp/${filename}-lf.tmp" "${1}"` when you should instead reassign the pointer to the new file in `/tmp`. So maybe something like 
```sh
# inside new shell script 
if [ "${exit_code}" == 0 ]; then
  echo "[info] '${1}' contains CRLF (Windows) line endings, converting to LF (Unix)..." | ts '%Y-%m-%d %H:%M:%.S'
  name="/tmp/${filename}-lf"
  rm -f "$name"
  dos2unix -n "${1}" "$name" 1> /dev/null
  echo "[info] Line ending conversion complete" | ts '%Y-%m-%d %H:%M:%.S'
  echo "$name"
fi
echo "$1"

# usage
VPN_CONFIG="$(dos2unix-wrapper.sh $VPN_CONFIG)"
```

Variable assignment described here https://stackoverflow.com/a/23565252/5156659

**Comment by binhex** [comment]
i was assuming the issue is in place editing of the file, the script i wrote does not do this, instead it uses dos2unix -n option to create a new file, once the new file is created its then moved from the temporary location to the original file location, unless a NFS mount does not permit files copied from one file system to another (sounds crazy if thats true) then i would assume this would work, can you give it a try if i create a test tagged image?

**Comment by the0neWhoKnocks** [comment]
Rather than downloading a 1GB image for testing I updated my script to use parts of your new file. The `mv` logs an error, but still moved the file (just without preserving the permissions). The further permission changes via `chown` and `chmod` completely fail and halt the execution due to permissions issues. This worked for me though
```sh
# convert file
dos2unix -n "${VPN_CONFIG}" /tmp/converted.ovpn
# delete source file
rm "${VPN_CONFIG}"
# copy over converted file
cp --no-preserve=mode /tmp/converted.ovpn "${VPN_CONFIG}"
```
I can't speak for the other `dos2unix` use cases you have, but the one above is the only issue I encountered and can test reliably.

Honestly, it still feels safer to not touch user's files, and instead just dump them in `/tmp` and change your pointers if conversions had to occur.

**Comment by binhex** [comment]
> Honestly, it still feels safer to not touch user's files, and instead just dump them in `/tmp` and change your pointers if conversions had to occur.

the thing that makes me uncomfortable with this is that it would mean the file would be converted every time, not ideal really, plus there are still the permissions issues to work around, as root user (account that will be doing the conversion) will be the owner of the file, causing potential issues when the file needs to be read by $PUID. There is also the security risk of copying the ovpn file inside the container (certain vpn providers include the authentication as a private key embeded in the ovpn file). As well as all of that there are other calls to dos2unix for different files, which does beg the question, why is it not causing an issue for those files, see link below (line 17):-
https://github.com/binhex/arch-delugevpn/blob/aeb11d7b75eac65ac0ea58ac7ca3731da6871256/run/nobody/watchdog.sh

as you might of seen i have a LOT of other repo's including other vpn variants which again use calls to dos2unix so i would like to get this as tight as possible and ideally have a script i can call for all use cases for all the images.

**Comment by binhex** [comment]
one quick thought can you please do a ls -al on the host for the volume mapping for /config, also can you post your docker run command here, i would be interested to see what values you have for PUID, PGID and UMASK.

**Comment by binhex** [comment]
just trying to replicate the issue here (so far i cant), and noticed with the version of dos2unix included in the image that the default behaviour is NOT to change file ownership, so im not sure why you are getting that message, taken from ```dos2unix --help```:-

``` --no-allow-chown      don't allow file ownership change (default)```

**Comment by the0neWhoKnocks** [comment]
Ok, so I started gathering all the data you requested - but before I submitted my reply I started digging through the NFS Service config within FreeNAS. Turns out I had `Enable NFSv4` checked, _but_ also needed `NFSv3 ownership model for NFSv4` checked. The second item apparently allows users/groups utilizing an NFS share to properly map UID/GID's. So even though `ls -la` was reporting the same user and group as what I was seeing within the Container there was some internal mapping that was being mis-matched causing the error.

Ugh, anyway, good news - you don't have to do anything 😃.  I'll close out the "issue". Thanks for your time.
